tensorflow-frameworks/tensorflow-data/README.md

```diff
@@ -0,0 +1,93 @@
+Tensorflow-Data (Java)
+==
+
+TensorFlow Data provides simple APIs for loading data of various formats, and preparing
+datasets for use in training and using deep learning models.
+
+TensorFlow Java's implementation simplifies the use of the C++ `data` ops, and provides
+a simple API for configuring and iterating over datasets in both "graph" and "eager" mode.
+
+Usage
+--
+
+The `Dataset` abstraction represents a sequence of elements, where each element in the sequence is a collection (`List`) of tensors (or, "components").
+
+
+Creation
+-
+A dataset can be constructed from a list of constant tensors
+using `Dataset.fromTensorSlices( ... )` as follows:
+
+```java
+// Declare dataset components as arrays.
+// NOTE: All components in a dataset must share the first "batch" dimension.
+
+int[][] m1 = new int[][]{
+        {1, 2, 3, 4, 5},
+        {2, 4, 6, 8, 10},
+        {3, 6, 8, 12, 15},
+        {4, 8, 12, 16, 20}
+    };
+
+int[][] m2 = new int[][]{
+    {1}, {0}, {1}, {1}
+};
+
+
+Ops tf = // ... TensorFlow Ops Accessor (either graph or eager).
+
+// Construct dataset with two components, batchSize=2.
+Dataset dataset = Dataset.fromTensorSlices(tf,
+    // List of array components
+    Arrays.asList(tf.constant(m1), tf.constant(m2)),
+    // List of each component's dtype
+    Arrays.asList(TInt32.DTYPE, TInt32.DTYPE)
+)
+```
+
+Iteration
+--
+
+In eager mode, the dataset can be iterated through using a standard 
+"for-each" loop, to receive the tensor values of each component:
+
+```java
+int BATCH_SIZE = 2;
+for (List<Output<?>> components : dataset.batch(BATCH_SIZE)) {
+      Tensor<TInt32> XBatch = components.get(0).tensor().expect(TInt32.DTYPE);
+      Tensor<TInt32> yBatch = components.get(1).tensor().expect(TInt32.DTYPE);
+      
+      // ... use batch tensors
+}
+```
+
+In graph mode, the dataset can be iterated through using the `OneShotIterator` abstraction, and a while loop, as follows:
+
+```java
+OneShotIterator oneShotIterator = dataset.makeOneShotIterator();
+Operation makeIterator = oneShotIterator.getMakeIteratorOp();
+List<Output<?>> components = oneShotIterator.getComponents();
+
+try (Session session = new Session(graph)) {
+    // Run MakeIterator Op to set iterator position
+    session.runner()
+        .addTarget(makeIterator)
+        .run();
+    
+    while (true) {
+        try {
+            List<Tensor<?>> outputs = session.runner()
+                .fetch(components.get(0))
+                .fetch(components.get(1))
+                .run();
+
+            Tensor<TInt32> matrix1 = outputs.get(0).expect(TInt32.DTYPE);
+            Tensor<TInt32> matrix2 = outputs.get(1).expect(TInt32.DTYPE);
+
+        } catch (IndexOutOfBoundsException e) {
```

Comment:
See the example here: https://www.tensorflow.org/api_docs/python/tf/compat/v1/data/Iterator#from_structure

I don't know of any way to avoid this nicely in graph mode. It doesn't really make sense to make it an `Iterable` since we don't actually want to fetch dataset batches at the beginning of the loop; we want to run the graph each iteration of the loop and get things like losses (all of which are constructed *before* opening a loop).

I think this API is ok for now; since it matches Python, it will be similar for users. Do you see any way of improving it from here?

Reply:
I agree that's how Python does it, but that's because Python's iteration idiom involves catching StopIteration when it's done, so the TF Python way mirrors that. Where does the `IndexOutOfBoundsException` come from? Wherever it's generated we could modify the code to set a flag on the session that could be used as a condition in the while loop. Or we could convert the output of `session.runner().run()` to be a results object that was closeable, and monitor a boolean set in that (that's got ugly scoping issues though). Or it could add a node which produces a scalar `Tensor` containing true or false that the user could monitor.

Those all assume there's only a single iterator on any given session (or a single root iterator), but I'm not sure from the code if multiple iterators are supported anyway. I'm not sure that any of these are particularly good solutions, but I really don't like catching runtime exceptions as the way users have to use the library.

Reply:
The `IndexOutOfBoundsException` is thrown when `session.run()` is called on a graph with dataset "getNext" components which are exhausted; so that'd not be easy to change, I think.

I think propagating the exception to the user is reasonable; hiding the exception can cause the user to misuse the abstraction (i.e. improperly tracking a boolean flag), and deal with undefined behavior...

Also, if the user already knows the size of the dataset (# batches), they can use a for-loop.

Multiple and re-initializeable iterators are supported; yup it does make things a little more complicated.

@karllessard what are your thoughts?

Reply:
But the exception must be thrown by some code we control, as the TF C library isn't going to throw `IndexOutOfBoundsException`. So if you can point me at the place where it's thrown then we can figure out what information is available then, and what we could do to make this behave like Java code.

Do you have an example somewhere of a Graph training loop with multiple epochs where we'd have to reinitialise the iterator? I've not seen one so far and I'd like to see how it's supposed to be used. Ditto for multiple iterators, as at the moment we won't know which iterator is exhausted, and AFAICT there's no way for the user to know that either?